Recorded future Feed Integration

reutshal opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Status
Ready

## Related Issues
reference: demisto/etc#19714

## Description
RecordedFuture feed with two sub-feeds

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [X] Code Review

## Technical writer review
Mention and link to the files that require a technical writer review.
- [ ] [YAML file](https://github.com/demisto/content/blob/recordedFuture-tip/TestPlaybooks/playbook-RecordedFutureFeed.yml)
- [ ] [CHANGELOG](link)

avidan-H commented on Integrations/FeedRecordedFuture/FeedRecordedFuture_test.py:
```diff
@@ -0,0 +1,80 @@
+import pytest
+from collections import OrderedDict
+from FeedRecordedFuture import get_indicator_type, split_hash_context, get_indicator_context
+
```
If indicator types is reduced to just `File`, then you will need to adjust the unit tests accordingly as well.

avidan-H commented:
@reutshal Also, please add a screenshot of the test playbook as well.

avidan-H commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -0,0 +1,331 @@
+BASE_URL = 'https://api.recordedfuture.com/v2/'
+PARAMS = {'output_format': 'csv/splunk'}
+HEADERS = {'X-RF-User-Agent': 'Demisto',
+           'content-type': 'application/json'}
```
I think it may make more sense to add these values as instance variables in your client's `__init__` method.

avidan-H commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.yml:
```diff
@@ -0,0 +1,185 @@
+      type: String
+    - contextPath: RecordedFutureFeed.Indicator.Value
+      description: The indicator value
+      type: Unknown
```
```suggestion
      type: String
```

reutshal commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -0,0 +1,331 @@
+    commands = {
+        'test-module': test_module,
+        'get-indicators': get_indicators_command,
+        'get-risk-rules': get_risk_rules_command
```
Done

reutshal commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.yml:
```diff
@@ -0,0 +1,185 @@
+  name: polling_timeout
+  required: true
+  type: 0
+description: This is recorded future feed integration
```
Done

reutshal commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.yml:
```diff
@@ -0,0 +1,185 @@
+    deprecated: false
+    description: Get indicators from the feed
+    execution: false
+    name: get-indicators
```
Done

reutshal commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.yml:
```diff
@@ -0,0 +1,185 @@
+    deprecated: false
+    description: Get a list of the risk rules available for an indicator
+    execution: false
+    name: get-risk-rules
```
Done

reutshal commented:
> @reutshal Also, please add a screenshot of the test playbook as well.

![image](https://user-images.githubusercontent.com/50294648/72791102-6991f200-3c3f-11ea-8f6d-f741dfe36957.png)

reutshal commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -0,0 +1,331 @@
+    else:
+        LOG(F'{INTEGRATION_NAME} - Unknown IP version: {address_type.version}')
+        return {}
+    return type_
```
Done

avidan-H commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -0,0 +1,266 @@
+        return FeedIndicatorType.URL
+
+
+def fetch_indicators_command(client, indicator_type):
```
```suggestion
def fetch_indicators_command(client, indicator_type, limit: Optional[int]):
```

reutshal replied:
Done

avidan-H commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -0,0 +1,266 @@
+    indicators_list = fetch_indicators_command(client, indicator_type)
+    entry_result = camelize(indicators_list[:limit])
+    hr = tableToMarkdown('Indicators from RecordedFuture Feed:', entry_result, headers=['Value', 'Type'])
+
```
Call `createIndicators` here as well.

ronykoz replied:
@anara123 @DeanArbel What do you think about this one?

avidan-H commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -0,0 +1,266 @@
+        """
+    indicator_type = args.get('indicator_type', demisto.params().get('indicator_type'))
+    limit = int(args.get('limit'))
+    indicators_list = fetch_indicators_command(client, indicator_type)
```
```suggestion
    indicators_list = fetch_indicators_command(client, indicator_type, limit)
```

reutshal replied:
Done

avidan-H commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -0,0 +1,266 @@
+    indicator_type = args.get('indicator_type', demisto.params().get('indicator_type'))
+    limit = int(args.get('limit'))
+    indicators_list = fetch_indicators_command(client, indicator_type)
+    entry_result = camelize(indicators_list[:limit])
```
```suggestion
    entry_result = camelize(indicators_list)
```

ronykoz commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -0,0 +1,289 @@
+        if demisto.command() == 'fetch-indicators':
+            indicators = fetch_indicators_command(client, client.indicator_type, None)
+            # we submit the indicators in batches
+            for b in old_batch(indicators, batch_size=2000):  # TODO change to commonserverpython batch
```
update to the CommonServerPython one

ronykoz commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -0,0 +1,289 @@
+        })
+    headers = ['Name', 'Description', 'Criticality']
+    hr = tableToMarkdown(f'Available risk rules for {indicator_type}:', entry_result, headers)
+    return hr, {'RecordedFutureFeed.RiskRule': entry_result}, result
```
```suggestion
    return hr, {'RecordedFutureFeed.RiskRule(val.Name == obj.Name)': entry_result}, result
```

ronykoz commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -0,0 +1,289 @@
+        return FeedIndicatorType.URL
+
+
+def ip_to_indicator_type(ip):
```
remove this once we have that in CommonServerPython

ronykoz commented on Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -0,0 +1,289 @@
+            yield current_batch
+            current_batch = []
+    if current_batch:
+        yield current_batch
```
remove

avidan-H commented on Scripts/CommonServerPython/CommonServerPython.yml:
```diff
@@ -16,4 +16,4 @@ timeout: 0s
 alt_dockerimages: # used for unit testing
```
Why is a change to CommonServerPython.yml file here? I thought you reverted the CommonServerPython related changes.

reutshal replied:
reverting

reutshal commented:
@kirbles19  could you please go over the yml file : 
https://github.com/demisto/content/blob/recordedFuture-tip/Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.yml
The parameters relevant to my integration:
Indicator type, API token, Sub Feed, Risk Rule, Fusion File Path.
Thanks!

ronykoz commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -146,7 +134,8 @@ def test_module(client: Client, args: dict) -> Tuple[str, dict, dict]:
     """
 
-    client.build_iterator('connectApi', 'ip')
+    for feed in client.sub_feed:
```
```suggestion
    for sub_feed in client.sub_feed:
```

avidan-H replied:
```suggestion
    for sub_feed in client.sub_feeds:
```

reutshal replied:
Done

kirbles19 commented:
@reutshal Done.

anara123 commented:
@reutshal don't forget that we need to add threshold and return score

reutshal commented:
> @reutshal don't forget that we need to add threshold and return score
Done.

reutshal commented:
@kirbles19 added the 'threshold' parameter to the yml, please take a look.
thanks!

reutshal commented:
@avidd @ronykoz this is a screenshot of the test-playbook:
![image](https://user-images.githubusercontent.com/50294648/73126027-3be9d780-3fb6-11ea-9121-e545563eda1c.png)

ronykoz commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -116,14 +118,35 @@ def build_iterator(self, sub_feed, indicator_type):
+        """
+        dbot_score = 0
+        risk_from_feed = int(risk_from_feed)
+        if risk_from_feed >= self.threshold or risk_from_feed >= 65:
```
why or risk_from_feed >= 65 ?

ronykoz commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -116,14 +118,35 @@ def build_iterator(self, sub_feed, indicator_type):
+        risk_from_feed = int(risk_from_feed)
+        if risk_from_feed >= self.threshold or risk_from_feed >= 65:
+            dbot_score = 3
+        elif risk_from_feed >= 5:
```
what does 5 mean?

reutshal replied:
![image](https://user-images.githubusercontent.com/50294648/73137214-77d87780-405e-11ea-90e5-eab661c58958.png)
this should explain the scoring.
after consulting @idovandijk , we decided to treat malicious+very malicious as malicious,
unusual+suspicious as suspicious

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.yml:
```diff
@@ -0,0 +1,186 @@
+  required: true
+  type: 0
+- additionalinfo: The minimum score from the feed in order to to determine whether
+    the indicator is malicious. Default is "65"
```
Perhaps include here a sentence or two about the scoring that Recorded Future uses as you showed us with your screenshot. Something along the lines of `Recorded Future's scoring system is out of 100. By default, Recorded Future deems indicators in the range blah blah blah to be Suspicious and blah blah blah Malicious etc.`.

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture_description.md:
```diff
@@ -0,0 +1,14 @@
+## Recorded Future Feed
+This integration downloads from Recorded Future a list of IP Addresses, domains, urls, or file hash with known risk. 
```
```suggestion
This integration downloads from Recorded Future a list of IP addresses, domains, URLs, or file hashes with known risk associations. 
```

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture_description.md:
```diff
@@ -0,0 +1,14 @@
+
+The feed can be configured to load a specific risk list by specifying a risk rule. 
+
+If no risklist is specified, Recorded Futures default risklist is used. 
```
```suggestion
If no risk list is specified, Recorded Future's default risk list is used. 
```

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture_description.md:
```diff
@@ -0,0 +1,14 @@
+
+If no risklist is specified, Recorded Futures default risklist is used. 
+
+The feed can also be configured to load a custom risklist from a user-specified Recorded Future file path.
```
```suggestion
The feed can also be configured to load a custom risk list from a user-specified Recorded Future file path.
```

avidan-H replied:
`The feed`? Which feed? Please specify.

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture_description.md:
```diff
@@ -0,0 +1,14 @@
+
+The feed can also be configured to load a custom risklist from a user-specified Recorded Future file path.
+
+If no file path is specified, the default risklist file is used. 
```
```suggestion
If no file path is specified, the default risk list file is used. 
```

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture_description.md:
```diff
@@ -0,0 +1,14 @@
+
+If no file path is specified, the default risklist file is used. 
+
+To access this resource, you need a valid Recorded Future API token.
```
```suggestion
To access this resource, a valid Recorded Future API token is required.
```

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.yml:
```diff
@@ -0,0 +1,181 @@
+configuration:
+- defaultvalue: A - Completely reliable
+  display: Source reliability
+  name: feedInstanceReliability
```
```suggestion
  name: feedReliability
```

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.yml:
```diff
@@ -0,0 +1,181 @@
+  type: 15
+- defaultvalue: suddenDeath
+  hidden: true
+  name: expirationPolicy
```
```suggestion
  name: feedExpirationPolicy
```

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.yml:
```diff
@@ -0,0 +1,181 @@
+  required: false
+  type: 0
+- hidden: true
+  name: expirationInterval
```
```suggestion
  name: feedExpirationInterval
  defaultvalue: "20160"
```

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.yml:
```diff
@@ -0,0 +1,181 @@
+  - E - Unreliable
+  - F - Reliability cannot be judged
+  required: true
+  type: 15
```
```suggestion
  type: 15
  additionalinfo: Reliability of the source providing the intelligence data
```

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.yml:
```diff
@@ -0,0 +1,181 @@
+    reputation.
+  defaultvalue: feedInstanceReputationNotSet
+  display: Indicator reputation
+  name: feedInstanceReputation
```
```suggestion
  name: feedReputation
```

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.yml:
```diff
@@ -0,0 +1,181 @@
+  display: Fetch Interval
+  name: feedFetchInterval
+  required: true
+  type: 19
```
You also need to include the `feedBypassExclusionList` parameter. Look at the confluence [page](https://confluence.paloaltonetworks.com/pages/viewpage.action?spaceKey=DemistoContent&title=Minemeld+miners+to+Demisto+feed+integration) towards the bottom to see the updated server parameter values. Add any you are missing, please. It might be best to follow the order given there as well.

reutshal replied:
Done

avidan-H commented on TestPlaybooks/playbook-RecordedFutureFeed.yml:
```diff
@@ -0,0 +1,669 @@
+id: RecordedFutureFeed - Test
```
Create a new directory in your pack called `TestPlaybooks`. Then, please move this file to that directory, and remove the `playbook-` prefix from the filename.

ronykoz commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -147,6 +154,44 @@ def calculate_indicator_score(self, risk_from_feed):
+                             "To see all available risk rules run the '!rf-get-risk-rules' command.")
+
+        if self.fusion_file_path is not None:
+            if 'fusion' not in self.sub_feeds:
```
you have another param which is for fusion only, don't you?

ronykoz replied:
risklist is for fusion though, we should add this check here

ronykoz replied:
so lets add this please

ronykoz commented:
lets add the traceback.format_exc() in the except located in the main( don't forget to import traceback before that)

And lets merge this thing

reutshal commented:
@kirbles19 Can you please go over the integration description I added:
https://github.com/demisto/content/blob/recordedFuture-tip/Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture_description.md
And also added the malicious threshold parameter to the yml:
https://github.com/demisto/content/blob/recordedFuture-tip/Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.yml
Thanks!

avidan-H commented on Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py:
```diff
@@ -188,7 +188,7 @@ def is_valid_risk_rule(client: Client, risk_rule):
         headers=client.headers
     )
-    risk_rules_list = [risk_rule['name'] for risk_rule in risk_rule_response['data']['results']]
+    risk_rules_list = [risk_rule['name'] for single_risk_rule in risk_rule_response['data']['results']]
```
??? This won't work.

avidan-H replied:
```suggestion
    risk_rules_list = [single_risk_rule['name'] for single_risk_rule in risk_rule_response['data']['results']]
```
Just don't use `risk_rule` in your list comprehension at all since that is the name of one of your function parameters here. If you decide to call it `single_risk_rule` then you need to change it there and before.

reutshal replied:
Of course, I forgot to also change the first one in the list comprehension.
